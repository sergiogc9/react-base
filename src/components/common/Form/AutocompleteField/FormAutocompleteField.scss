@import "@src/scss/_globals.scss";
@import "@src/scss/_mixins.scss";
@import "@src/scss/_variables.scss";

.react-base-form-autocomplete-field {
	@include selectInput("darker");
	label {
		font-size: 1.2rem;
	}

	hr.md-divider {
		margin-top: 0.4rem;
	}

	ul.md-list {
		padding-bottom: 0;
		li.md-list-item {
			> div {
				padding: 1rem 2rem;
				.md-text {
					font-size: 1.2rem;
				}
			}
		}
	}

	@include autocompleterWithClearButton();
	.media-option-name {
		font-weight: bold;
	}

	& + button {
		position: absolute;
		right: 2rem;
		bottom: 1rem;
	}

	&.with-button:focus-within input {
		padding-right: 2rem;
	}
}
