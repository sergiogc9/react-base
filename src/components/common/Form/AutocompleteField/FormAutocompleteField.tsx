import React from 'react';
import Autocomplete from 'react-md/lib/Autocompletes/Autocomplete';
import Button from 'react-md/lib/Buttons';
import isURL from 'validator/lib/isURL';
import get from 'lodash/get';
import has from 'lodash/has';

import { withT, TProps } from '@src/lib/i18n';
import { AutocompleteFieldElement, FormFieldItem } from '@src/types/form';
import { formFieldBaseValidation } from '@src/components/common/Form/Form';
import { useTimeout, useForceFieldValue } from '@src/lib/hooks';

import './FormAutocompleteField.scss';

export type ComponentProps = TProps & {
	element: AutocompleteFieldElement,
	forceValue?: FormFieldItem,
	error: boolean,
	onAutocompleteSelected: (elementId: string, value: string, isValid: boolean) => void,
	onUpdateItems?: (newValue: string) => Promise<AutocompleteFieldElement["items"]>
};

const __validateAutocompleteField = (element: AutocompleteFieldElement, newItem: FormFieldItem) => {
	if (!formFieldBaseValidation(element, newItem.value)) return false;
	if (!element.validations) return true;
	if (element.validations.url && !isURL(newItem.value)) return false;
	return true;
};

const __mapItems = (items: AutocompleteFieldElement["items"]) => items.map(item => ({ label: item.label, value: item.value, key: item.value }));

const FormAutocompleteField = (props: ComponentProps) => {
	const { t, element, forceValue, error, onAutocompleteSelected, onUpdateItems } = props;

	const defaultItems = React.useMemo(() => __mapItems(element.items), [element.items]);
	const [selectedItem, setSelectedItem] = React.useState<FormFieldItem>(element.defaultValue ? element.defaultValue : { label: "", value: "" });
	const [items, setItems] = React.useState(defaultItems);
	const { run: runTO, clear: clearTO } = useTimeout();
	useForceFieldValue(forceValue, v => v.value, element, setSelectedItem, onAutocompleteSelected, __validateAutocompleteField);

	React.useEffect(() => {
		if (onUpdateItems && selectedItem.value) {
			const updateItems = async (val: string) => {
				setItems(__mapItems(await onUpdateItems(val)));
			};

			runTO(() => updateItems(selectedItem.value), 200);
		}
		return clearTO;
	}, [selectedItem.value, onUpdateItems, runTO, clearTO]);

	const __onChangeHandler = React.useCallback((val: string) => setSelectedItem({ label: val, value: val }), []);
	const __onBlurHandler = React.useCallback(
		() => onAutocompleteSelected(element.id, selectedItem.value, __validateAutocompleteField(element, selectedItem)),
		[element, selectedItem, onAutocompleteSelected]
	);
	const __onAutocompleteHandler = React.useCallback((suggestion, index, results: any) => setSelectedItem({
		label: element.useValueFromAutocomplete ? results[index].value : results[index].label,
		value: results[index].value
	}), [element]);
	const __onCloseBtnClickHandler = React.useCallback(() => setSelectedItem({ label: "", value: "" }), []);

	return (
		<>
			<Autocomplete
				id={`formAutocompleteField-${element.id}`}
				className={`discover-form-autocomplete-field ${element.closeButton ? "with-button " : ""}${element.className || ""}`}
				label={has(element, 'label.key') ? t(get(element, 'label.key')) : element.label}
				data={items}
				dataLabel="label"
				dataValue="value"
				value={selectedItem.label}
				onChange={__onChangeHandler}
				onBlur={__onBlurHandler}
				onAutocomplete={__onAutocompleteHandler}
				error={error}
				showUnfilteredData={element.alwaysShowItems}
				clearOnAutocomplete={false}
				{...element.inputProps}
			/>
			{element.closeButton ?
				<Button
					id={`formAutocompleteFieldCloseBtn-${element.id}`}
					icon
					iconChildren="close"
					children=""
					onChange={__onCloseBtnClickHandler}
				/> : null}

		</>
	);
};

export default withT(React.memo(FormAutocompleteField));
