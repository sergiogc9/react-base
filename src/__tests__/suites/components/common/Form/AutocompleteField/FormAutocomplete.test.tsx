import React from "react";
import { mount } from "enzyme";
import { act } from 'react-dom/test-utils';

import FormAutocompleteField from "@src/components/common/Form/AutocompleteField/FormAutocompleteField";
import { ComponentProps } from "@src/components/common/Form/AutocompleteField/FormAutocompleteField";
import TestHelper from "@src/__tests__/utils/Helper/TestHelper";

const element = TestHelper.autocompleteFieldElement;

const items = TestHelper.formFieldItems;

jest.useFakeTimers();
let onChangeMock = jest.fn();
let onUpdateItemsMock = jest.fn();
describe("Autocomplete field common component", () => {
	let wrapper: any;

	const updateComponent = (props: Partial<ComponentProps> = {}) => {
		act(() => {
			wrapper = mount(
				<FormAutocompleteField
					element={element}
					error={false}
					onAutocompleteSelected={onChangeMock}
					{...props}
				/>);
		});
	};

	beforeEach(() => {
		onChangeMock = jest.fn();
		onUpdateItemsMock = jest.fn();
		updateComponent();
	});

	it("Component default snapshot", () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot with label with key", () => {
		updateComponent({ element: { ...element, label: { key: "lang_key" } } });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot with close button", () => {
		updateComponent({ element: { ...element, closeButton: true } });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot with items", () => {
		updateComponent({ element: { ...element, items } });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component on update items handler called", () => {
		onUpdateItemsMock.mockReturnValueOnce(new Promise(() => items));
		updateComponent({ element: { ...element }, onUpdateItems: onUpdateItemsMock });
		act(() => wrapper.find("Autocomplete").at(0).props().onChange("newVal"));
		act(() => { jest.runAllTimers(); });
		expect(onUpdateItemsMock).toHaveBeenCalledTimes(1);
	});

	it("Component snapshot after on change value", () => {
		act(() => wrapper.find("Autocomplete").at(0).props().onChange("newVal"));
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component on blur event", () => {
		act(() => wrapper.find("Autocomplete").at(0).props().onBlur());
		expect(onChangeMock).toHaveBeenCalledTimes(1);
	});

	it("Component snapshot after on autocomplete event", () => {
		act(() => wrapper.find("Autocomplete").at(0).props().onAutocomplete(null, 0, [{ label: "newlab", value: "newval" }]));
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot after on autocomplete event with useValueFromAutocomplete flag enabled", () => {
		wrapper.setProps({ element: { ...element, useValueFromAutocomplete: true } });
		act(() => wrapper.find("Autocomplete").at(0).props().onAutocomplete(null, 0, [{ label: "newlab", value: "newval" }]));
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component snapshot after on close event click handler", () => {
		updateComponent({ element: { ...element, closeButton: true, defaultValue: { label: "labb", value: "vall" } } });
		act(() => wrapper.find("Button").at(0).props().onChange());
		expect(wrapper.html()).toMatchSnapshot();
	});

	it("Component on validate base validation", () => {
		updateComponent({ element: { ...element, defaultValue: { label: "labb", value: "vall" }, validations: { fn: () => false } } });
		act(() => wrapper.find("Autocomplete").at(0).props().onBlur());
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "vall", false);
	});

	it("Component on validate url", () => {
		updateComponent({ element: { ...element, defaultValue: { label: "labb", value: "vall" }, validations: { url: true } } });
		act(() => wrapper.find("Autocomplete").at(0).props().onBlur({ target: { value: 15 } }));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "vall", false);
	});

	it("Component on validate correct", () => {
		updateComponent({ element: { ...element, defaultValue: { label: "labb", value: "https://gironafc.cat" }, validations: { url: true } } });
		act(() => wrapper.find("TextField").at(0).props().onBlur({ target: { value: 7 } }));
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "https://gironafc.cat", true);
	});

	it("Component snapshot after on force update", () => {
		wrapper.setProps({ forceValue: { label: "labnew", value: "valnew" } });
		expect(onChangeMock).toHaveBeenCalledWith(element.id, "valnew", true);
	});
});
