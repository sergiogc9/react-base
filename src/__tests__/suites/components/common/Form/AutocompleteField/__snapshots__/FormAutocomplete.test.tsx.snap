// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete field common component Component default snapshot 1`] = `"<div id=\\"formAutocompleteField-autocompleteFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu-container--menu-below md-autocomplete-container react-base-form-autocomplete-field \\"><div class=\\"md-text-field-container md-full-width md-text-field-container--input md-autocomplete\\"><label for=\\"formAutocompleteField-autocompleteFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">Text label</label><input autocomplete=\\"off\\" aria-autocomplete=\\"list\\" id=\\"formAutocompleteField-autocompleteFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div></div>"`;

exports[`Autocomplete field common component Component snapshot after on autocomplete event 1`] = `"<div id=\\"formAutocompleteField-autocompleteFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu-container--menu-below md-autocomplete-container react-base-form-autocomplete-field \\"><div class=\\"md-text-field-container md-full-width md-text-field-container--input md-autocomplete\\"><label for=\\"formAutocompleteField-autocompleteFieldId\\" class=\\"md-floating-label md-floating-label--floating md-text--secondary\\">Text label</label><input autocomplete=\\"off\\" aria-autocomplete=\\"list\\" id=\\"formAutocompleteField-autocompleteFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"newlab\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div></div>"`;

exports[`Autocomplete field common component Component snapshot after on autocomplete event with useValueFromAutocomplete flag enabled 1`] = `"<div id=\\"formAutocompleteField-autocompleteFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu-container--menu-below md-autocomplete-container react-base-form-autocomplete-field \\"><div class=\\"md-text-field-container md-full-width md-text-field-container--input md-autocomplete\\"><label for=\\"formAutocompleteField-autocompleteFieldId\\" class=\\"md-floating-label md-floating-label--floating md-text--secondary\\">Text label</label><input autocomplete=\\"off\\" aria-autocomplete=\\"list\\" id=\\"formAutocompleteField-autocompleteFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"newval\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div></div>"`;

exports[`Autocomplete field common component Component snapshot after on change value 1`] = `"<div id=\\"formAutocompleteField-autocompleteFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu-container--menu-below md-autocomplete-container react-base-form-autocomplete-field \\"><div class=\\"md-text-field-container md-full-width md-text-field-container--input md-autocomplete\\"><label for=\\"formAutocompleteField-autocompleteFieldId\\" class=\\"md-floating-label md-floating-label--floating md-text--secondary\\">Text label</label><input autocomplete=\\"off\\" aria-autocomplete=\\"list\\" id=\\"formAutocompleteField-autocompleteFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"newVal\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div></div>"`;

exports[`Autocomplete field common component Component snapshot after on close event click handler 1`] = `"<div id=\\"formAutocompleteField-autocompleteFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu-container--menu-below md-autocomplete-container react-base-form-autocomplete-field with-button \\"><div class=\\"md-text-field-container md-full-width md-text-field-container--input md-autocomplete\\"><label for=\\"formAutocompleteField-autocompleteFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">Text label</label><input autocomplete=\\"off\\" aria-autocomplete=\\"list\\" id=\\"formAutocompleteField-autocompleteFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div></div><button id=\\"formAutocompleteFieldCloseBtn-autocompleteFieldId\\" type=\\"button\\" class=\\"md-btn md-btn--icon md-pointer--hover md-inline-block\\"><div class=\\"md-ink-container\\"></div><i class=\\"md-icon material-icons md-text--inherit\\">close</i></button>"`;

exports[`Autocomplete field common component Component snapshot with close button 1`] = `"<div id=\\"formAutocompleteField-autocompleteFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu-container--menu-below md-autocomplete-container react-base-form-autocomplete-field with-button \\"><div class=\\"md-text-field-container md-full-width md-text-field-container--input md-autocomplete\\"><label for=\\"formAutocompleteField-autocompleteFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">Text label</label><input autocomplete=\\"off\\" aria-autocomplete=\\"list\\" id=\\"formAutocompleteField-autocompleteFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div></div><button id=\\"formAutocompleteFieldCloseBtn-autocompleteFieldId\\" type=\\"button\\" class=\\"md-btn md-btn--icon md-pointer--hover md-inline-block\\"><div class=\\"md-ink-container\\"></div><i class=\\"md-icon material-icons md-text--inherit\\">close</i></button>"`;

exports[`Autocomplete field common component Component snapshot with items 1`] = `"<div id=\\"formAutocompleteField-autocompleteFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu-container--menu-below md-autocomplete-container react-base-form-autocomplete-field \\"><div class=\\"md-text-field-container md-full-width md-text-field-container--input md-autocomplete\\"><label for=\\"formAutocompleteField-autocompleteFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">Text label</label><input autocomplete=\\"off\\" aria-autocomplete=\\"list\\" id=\\"formAutocompleteField-autocompleteFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div></div>"`;

exports[`Autocomplete field common component Component snapshot with label with key 1`] = `"<div id=\\"formAutocompleteField-autocompleteFieldId-menu\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"md-layover md-layover--simplified md-full-width md-menu-container md-menu-container--menu-below md-autocomplete-container react-base-form-autocomplete-field \\"><div class=\\"md-text-field-container md-full-width md-text-field-container--input md-autocomplete\\"><label for=\\"formAutocompleteField-autocompleteFieldId\\" class=\\"md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--secondary\\">lang_key</label><input autocomplete=\\"off\\" aria-autocomplete=\\"list\\" id=\\"formAutocompleteField-autocompleteFieldId\\" type=\\"text\\" class=\\"md-text-field md-text-field--floating-margin md-full-width md-text\\" value=\\"\\"><hr class=\\"md-divider md-divider--text-field md-divider--expand-from-left\\"></div></div>"`;
